icu4j/main/classes/core/src/com/ibm/icu/text/RuleBasedBreakIterator.java

```diff
@@ -1436,42 +1436,84 @@ int current() {
     boolean populateNear(int position) {
         assert(position < fBoundaries[fStartBufIdx] || position > fBoundaries[fEndBufIdx]);
 
-        // Find a boundary somewhere in the vicinity of the requested position.
-        // Depending on the safe rules and the text data, it could be either before, at, or after
-        // the requested position.
-
+        // Add boundaries to the cache near the specified position.
+        // The given position need not be a boundary itself.
+        // The input position must be within the range of the text, and
+        // on a code point boundary.
+        // If the requested position is a break boundary, leave the iteration
+        // position on it.
+        // If the requested position is not a boundary, leave the iteration
+        // position on the preceding boundary and include both the
+        // preceding and following boundaries in the cache.
+        // Additional boundaries, either preceding or following, may be added
+        // to the cache as a side effect.
 
         // If the requested position is not near already cached positions, clear the existing cache,
         // find a near-by boundary and begin new cache contents there.
 
-        if ((position < fBoundaries[fStartBufIdx] - 15) || position > (fBoundaries[fEndBufIdx] + 15)) {
-            int aBoundary = fText.getBeginIndex();
-            int ruleStatusIndex = 0;
-            if (position > aBoundary + 20) {
-                int backupPos = handleSafePrevious(position);
-                if (backupPos > aBoundary) {
-                    // Advance to the boundary following the backup position.
-                    // There is a complication: the safe reverse rules identify pairs of code points
-                    // that are safe. If advancing from the safe point moves forwards by less than
-                    // two code points, we need to advance one more time to ensure that the boundary
-                    // is good, including a correct rules status value.
-                    //
-                    fPosition = backupPos;
+        int startOfText = fText.getBeginIndex();
+        int aBoundary = -1;
+        int ruleStatusIndex = 0;
+        boolean retainCache = false;
+        if ((position > fBoundaries[fStartBufIdx] - 15) && position < (fBoundaries[fEndBufIdx] + 15)) {
+            // Requested position is near the existing cache. Retain it.
+            retainCache = true;
+        } else if (position <= startOfText + 20) {
```

Comment:
what is the magic number 20 here? Could we add a constant for that and add a note about how we pick that number?